docarray/array/storage/memory/backend.py

```diff
@@ -40,26 +54,32 @@ def _rebuild_id2offset(self) -> None:
             d.id: i for i, d in enumerate(self._data)
         }  # type: Dict[str, int]
 
+        self._needs_id2offset_rebuild = False
+
+    @needs_id2offset_rebuild
     def _init_storage(
         self, _docs: Optional['DocumentArraySourceType'] = None, copy: bool = False
     ):
         from ... import DocumentArray
+        from ...memory import DocumentArrayInMemory
 
         self._data = []
+        self._id_to_index = {}
         if _docs is None:
             return
         elif isinstance(
             _docs, (DocumentArray, Sequence, Generator, Iterator, itertools.chain)
         ):
             if copy:
                 self._data = [Document(d, copy=True) for d in _docs]
-                self._rebuild_id2offset()
             elif isinstance(_docs, DocumentArray):
                 self._data = _docs._data
                 self._id_to_index = _docs._id2offset
```

Comment:
i think this part is wrong, not all `DocumentArray` has `_id_to_index` or `_id2offset`

Reply:
well either the DocumentArray inherit from BackendMixin and has `_id_to_index` or `_id2offset` (because it is defined in the this class), either they don't inherit and they don't call this method. So in both case everything is fine ? Might be missing something

Reply:
I agree, that was here before so I might be an other bug ? we should delete line 77 as the index _id_to_index will be automatically built on first call.

Reply:
fix in last commit